In [1]:
df = pd.read_csv("data_one_per_bit_largest_full.csv")

<IPython.core.display.Javascript object>

In [2]:
df.head()

real_1  real_2  real_3  real_4  real_5  ...  imag_998  imag_999  imag_1000  SMF_length  Type_of_modulation
0  0.0019  0.0206  0.0114  0.0078  0.0383  ...   -0.0018   -0.0003     0.0009          10                 OOK
1  0.0203  0.0139  0.0168  0.0158  0.0207  ...   -0.0046    0.0001    -0.0017          10                 OOK
2  0.0146  0.0118  0.0343 -0.0007  0.0132  ...   -0.0068    0.0001     0.0022          10                 OOK
3 -0.0034  0.0356  0.0350 -0.0076  0.0314  ...   -0.0063   -0.0001    -0.0019          10                 OOK
4  0.0260  0.0377  0.0246 -0.0121  0.0415  ...   -0.0037    0.0001     0.0000          10                 OOK

[5 rows x 2002 columns]

In [3]:
modulation_type = pd.get_dummies(df['Type_of_modulation'])
modulation_type

<IPython.core.display.Javascript object>

BPSK  OOK  QPSK
0         0    1     0
1         0    1     0
2         0    1     0
3         0    1     0
4         0    1     0
...     ...  ...   ...
34795     0    0     1
34796     0    0     1
34797     0    0     1
34798     0    0     1
34799     0    0     1

[34800 rows x 3 columns]

In [4]:
df.drop('Type_of_modulation', axis=1, inplace=True)

In [5]:
df.head()

real_1  real_2  real_3  real_4  real_5  real_6  ...  imag_996  imag_997  imag_998  imag_999  imag_1000  SMF_length
0  0.0019  0.0206  0.0114  0.0078  0.0383  0.0133  ...       0.0   -0.0009   -0.0018   -0.0003     0.0009          10
1  0.0203  0.0139  0.0168  0.0158  0.0207  0.0174  ...      -0.0    0.0016   -0.0046    0.0001    -0.0017          10
2  0.0146  0.0118  0.0343 -0.0007  0.0132  0.0263  ...       0.0   -0.0022   -0.0068    0.0001     0.0022          10
3 -0.0034  0.0356  0.0350 -0.0076  0.0314  0.0410  ...      -0.0    0.0018   -0.0063   -0.0001    -0.0019          10
4  0.0260  0.0377  0.0246 -0.0121  0.0415  0.0223  ...       0.0    0.0001   -0.0037    0.0001     0.0000          10

[5 rows x 2001 columns]

In [6]:
df = pd.concat([df, modulation_type], axis = 1)


<IPython.core.display.Javascript object>

In [7]:
df.head()

real_1  real_2  real_3  real_4  real_5  real_6  real_7  ...  imag_998  imag_999  imag_1000  SMF_length  BPSK  OOK  QPSK
0  0.0019  0.0206  0.0114  0.0078  0.0383  0.0133  0.0244  ...   -0.0018   -0.0003     0.0009          10     0    1     0
1  0.0203  0.0139  0.0168  0.0158  0.0207  0.0174  0.0350  ...   -0.0046    0.0001    -0.0017          10     0    1     0
2  0.0146  0.0118  0.0343 -0.0007  0.0132  0.0263  0.0525  ...   -0.0068    0.0001     0.0022          10     0    1     0
3 -0.0034  0.0356  0.0350 -0.0076  0.0314  0.0410  0.0194  ...   -0.0063   -0.0001    -0.0019          10     0    1     0
4  0.0260  0.0377  0.0246 -0.0121  0.0415  0.0223  0.0210  ...   -0.0037    0.0001     0.0000          10     0    1     0

[5 rows x 2004 columns]

In [8]:
# output_categories = pd.concat([df['SMF_length'], df['BPSK'], df['OOK'], df['QPSK']], axis = 1)
output_categories = pd.concat([df['BPSK'], df['OOK'], df['QPSK']], axis = 1)
# output_categories = pd.concat([df['OOK'], df['QPSK']], axis = 1)
output_categories.shape

<IPython.core.display.Javascript object>

(34800, 3)

In [9]:
output_categories

BPSK  OOK  QPSK
0         0    1     0
1         0    1     0
2         0    1     0
3         0    1     0
4         0    1     0
...     ...  ...   ...
34795     0    0     1
34796     0    0     1
34797     0    0     1
34798     0    0     1
34799     0    0     1

[34800 rows x 3 columns]

In [10]:
x_train, x_test, y_train, y_test = train_test_split(df.drop(['SMF_length', 'BPSK', 'OOK', 'QPSK'], axis = 1),
                                                output_categories,
                                                test_size = 0.3,
                                                random_state = 101)
# x_train, x_test, y_train, y_test = train_test_split(df.drop(['SMF_length', 'OOK', 'QPSK'], axis = 1),
#                                                 output_categories,
#                                                 test_size = 0.3,
#                                                 random_state = 101)

C:\Users\Yuvi\Anaconda3\lib\importlib\_bootstrap.py:219: RuntimeWarning: numpy.ufunc size changed, may indicate binary incompatibility. Expected 192 from C header, got 216 from PyObject
  return f(*args, **kwds)
C:\Users\Yuvi\Anaconda3\lib\importlib\_bootstrap.py:219: RuntimeWarning: numpy.ufunc size changed, may indicate binary incompatibility. Expected 192 from C header, got 216 from PyObject
  return f(*args, **kwds)
C:\Users\Yuvi\Anaconda3\lib\importlib\_bootstrap.py:219: RuntimeWarning: numpy.ufunc size changed, may indicate binary incompatibility. Expected 192 from C header, got 216 from PyObject
  return f(*args, **kwds)
C:\Users\Yuvi\Anaconda3\lib\importlib\_bootstrap.py:219: RuntimeWarning: numpy.ufunc size changed, may indicate binary incompatibility. Expected 192 from C header, got 216 from PyObject
  return f(*args, **kwds)
C:\Users\Yuvi\Anaconda3\lib\importlib\_bootstrap.py:219: RuntimeWarning: numpy.ufunc size changed, may indicate binary incompatibility. Expected 192 fro

<IPython.core.display.Javascript object>

In [11]:
X_train, x_val, Y_train, y_val = train_test_split(df.drop(['SMF_length', 'BPSK', 'OOK', 'QPSK'], axis = 1),
                                                output_categories,
                                                test_size = 0.3,
                                                random_state = 101)
# X_train, x_val, Y_train, y_val = train_test_split(df.drop(['SMF_length', 'OOK', 'QPSK'], axis = 1),
#                                                 output_categories,
#                                                 test_size = 0.3,
#                                                 random_state = 101)

<IPython.core.display.Javascript object>

In [12]:
import tensorflow as tf
from tensorflow import *

In [13]:
x_train.shape[1]

2000

In [14]:
x_train = np.array(x_train).reshape(-1, x_train.shape[1], 1)
x_test = np.array(x_test).reshape(-1,  x_test.shape[1], 1)
y_train = np.array(y_train).reshape(-1, y_train.shape[1])
y_test = np.array(y_test).reshape(-1, y_test.shape[1])
x_val = np.array(x_val).reshape(-1,  x_val.shape[1], 1)
y_val = np.array(y_val).reshape(-1, y_val.shape[1])
x_train.shape

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

(24360, 2000, 1)

In [15]:
y_val.shape

(10440, 3)

In [16]:
x_train.shape[1:]
# y_train.shape

(2000, 1)

In [17]:
from tensorflow.keras import callbacks 
earlystopping = callbacks.EarlyStopping(monitor ="val_loss",  
                                        mode ="min", patience = 5,  
                                        restore_best_weights = True) 

In [18]:
model_CNN_simple = tf.keras.Sequential([
        tf.keras.layers.Conv1D(32,kernel_size = 3, activation = 'relu', input_shape = x_train.shape[1:]),
        tf.keras.layers.MaxPooling1D(pool_size=3),
        tf.keras.layers.Dropout(0.2),
        tf.keras.layers.Conv1D(64,kernel_size = 3, activation = 'relu'),
        tf.keras.layers.MaxPooling1D(pool_size=3),
        tf.keras.layers.Dropout(0.2),
        tf.keras.layers.Conv1D(64,kernel_size = 3, activation = 'relu'),
        tf.keras.layers.MaxPooling1D(pool_size=3),
        tf.keras.layers.Dropout(0.2),
        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(1024, activation = 'relu'),
        tf.keras.layers.Dense(512, activation = 'relu'),
        tf.keras.layers.Dense(3, activation = 'softmax')
        ])

In [19]:
model_CNN_simple.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv1d (Conv1D)              (None, 1998, 32)          128       
_________________________________________________________________
max_pooling1d (MaxPooling1D) (None, 666, 32)           0         
_________________________________________________________________
dropout (Dropout)            (None, 666, 32)           0         
_________________________________________________________________
conv1d_1 (Conv1D)            (None, 664, 64)           6208      
_________________________________________________________________
max_pooling1d_1 (MaxPooling1 (None, 221, 64)           0         
_________________________________________________________________
dropout_1 (Dropout)          (None, 221, 64)           0         
_________________________________________________________________
conv1d_2 (Conv1D)            (None, 219, 64)           1

In [20]:
model_CNN_simple.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

In [21]:
model_CNN_simple.fit(x_train, y_train, batch_size=64, verbose=1, epochs=50,validation_data=(x_val, y_val))
# model_CNN_simple.fit(x_train, y_train, batch_size=64, verbose=1, epochs=50,validation_data=(x_val, y_val), callbacks = [earlystopping])
# model_CNN_simple.fit(x_train, y_train, batch_size=32, epochs=5, verbose=1)

Train on 24360 samples, validate on 10440 samples
Epoch 1/50
24360/24360 [==============================] - 80s 3ms/sample - loss: 0.5293 - accuracy: 0.7606 - val_loss: 0.2908 - val_accuracy: 0.8775
Epoch 2/50
24360/24360 [==============================] - 71s 3ms/sample - loss: 0.2373 - accuracy: 0.9062 - val_loss: 0.1234 - val_accuracy: 0.9507
Epoch 3/50
24360/24360 [==============================] - 78s 3ms/sample - loss: 0.2372 - accuracy: 0.9079 - val_loss: 0.1511 - val_accuracy: 0.9470
Epoch 4/50
24360/24360 [==============================] - 85s 3ms/sample - loss: 0.1465 - accuracy: 0.9444 - val_loss: 0.0968 - val_accuracy: 0.9644
Epoch 5/50
24360/24360 [==============================] - 80s 3ms/sample - loss: 0.1110 - accuracy: 0.9594 - val_loss: 0.1265 - val_accuracy: 0.9488
Epoch 6/50
24360/24360 [==============================] - 78s 3ms/sample - loss: 0.1008 - accuracy: 0.9615 - val_loss: 0.0734 - val_accuracy: 0.9719
Epoch 7/50
24360/24360 [==============================] 

In [23]:
model_CNN_simple.evaluate(x_test, y_test, verbose=1)

10440/1 [===============================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================

[0.16995942146995346, 0.95019156]

In [24]:
for i in range(1, 10):
    x_val[i]
    y_val[i]
    testing_x = np.array(x_val[i]).reshape(-1, x_val.shape[1], 1)
    testing_y = np.array(y_val[i]).reshape(-1, y_val.shape[1])
    x = model_CNN_simple.predict(testing_x).round(1)
    print(f"Predicted {x}, Expected : {testing_y}")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[0.1 0.9 0. ]], Expected : [[1 0 0]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[0. 1. 0.]], Expected : [[0 1 0]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[0. 0. 1.]], Expected : [[0 0 1]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[0. 1. 0.]], Expected : [[0 1 0]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[0. 1. 0.]], Expected : [[0 1 0]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[0. 0. 1.]], Expected : [[0 0 1]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[0.9 0.1 0. ]], Expected : [[1 0 0]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[0. 1. 0.]], Expected : [[0 1 0]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[0. 0. 1.]], Expected : [[0 0 1]]


In [1]:
df = pd.read_csv("data_one_per_bit_largest_full.csv")

<IPython.core.display.Javascript object>

In [2]:
df.drop(['Type_of_modulation'], axis=1, inplace = True)

In [3]:
x_train, x_test, y_train, y_test = train_test_split(df.drop(['SMF_length'], axis = 1),
                                                df['SMF_length'],
                                                test_size = 0.3,
                                                random_state = 101)

<IPython.core.display.Javascript object>

In [4]:
X_train, x_val, Y_train, y_val = train_test_split(df.drop(['SMF_length'], axis = 1),
                                                df['SMF_length'],
                                                test_size = 0.3,
                                                random_state = 101)

<IPython.core.display.Javascript object>

In [5]:
y_train.shape[0]

24360

In [6]:
x_train = np.array(x_train).reshape(-1, x_train.shape[1], 1)
x_test = np.array(x_test).reshape(-1,  x_test.shape[1], 1)
y_train = np.array(y_train).reshape(y_train.shape[0], 1)
y_test = np.array(y_test).reshape(y_test.shape[0], 1)
x_val = np.array(x_val).reshape(-1,  x_val.shape[1], 1)
y_val = np.array(y_val).reshape(y_val.shape[0], 1)
x_train.shape

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

(24360, 2000, 1)

In [7]:
y_test.shape

(10440, 1)

In [8]:
model_length = tf.keras.Sequential([
        tf.keras.layers.Conv1D(32,kernel_size = 3, activation = 'relu', input_shape = x_train.shape[1:]),
        tf.keras.layers.MaxPooling1D(pool_size=3),
        tf.keras.layers.Dropout(0.2),
        tf.keras.layers.Conv1D(64,kernel_size = 3, activation = 'relu'),
        tf.keras.layers.MaxPooling1D(pool_size=3),
        tf.keras.layers.Dropout(0.2),
        tf.keras.layers.Conv1D(64,kernel_size = 3, activation = 'relu'),
        tf.keras.layers.MaxPooling1D(pool_size=3),
        tf.keras.layers.Dropout(0.2),
        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(1024, activation = 'relu'),
        tf.keras.layers.Dense(512, activation = 'relu'),
        tf.keras.layers.Dense(1, activation = 'relu')
        ])

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [9]:
model_length.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv1d (Conv1D)              (None, 1998, 32)          128       
_________________________________________________________________
max_pooling1d (MaxPooling1D) (None, 666, 32)           0         
_________________________________________________________________
dropout (Dropout)            (None, 666, 32)           0         
_________________________________________________________________
conv1d_1 (Conv1D)            (None, 664, 64)           6208      
_________________________________________________________________
max_pooling1d_1 (MaxPooling1 (None, 221, 64)           0         
_________________________________________________________________
dropout_1 (Dropout)          (None, 221, 64)           0         
_________________________________________________________________
conv1d_2 (Conv1D)            (None, 219, 64)           1

In [10]:
model_length.compile(loss='mae', optimizer='adam')

In [11]:
# model_length.fit(x_train, y_train, batch_size=64, verbose=1, epochs=50,validation_data=(x_val, y_val), callbacks = [earlystopping])
model_length.fit(x_train, y_train, batch_size=64, verbose=1, epochs=50,validation_data=(x_val, y_val))

Train on 24360 samples, validate on 10440 samples
Epoch 1/50
24360/24360 [==============================] - 104s 4ms/sample - loss: 19.7243 - val_loss: 19.0233
Epoch 2/50
24360/24360 [==============================] - 101s 4ms/sample - loss: 11.3487 - val_loss: 11.5551
Epoch 3/50
24360/24360 [==============================] - 102s 4ms/sample - loss: 10.5787 - val_loss: 10.9011
Epoch 4/50
24360/24360 [==============================] - 100s 4ms/sample - loss: 9.5156 - val_loss: 8.5437
Epoch 5/50
24360/24360 [==============================] - 113s 5ms/sample - loss: 8.8151 - val_loss: 9.2101
Epoch 6/50
24360/24360 [==============================] - 121s 5ms/sample - loss: 8.4935 - val_loss: 7.6561
Epoch 7/50
24360/24360 [==============================] - 101s 4ms/sample - loss: 8.5969 - val_loss: 8.6606
Epoch 8/50
24360/24360 [==============================] - 101s 4ms/sample - loss: 8.2519 - val_loss: 8.3206
Epoch 9/50
24360/24360 [==============================] - 101s 4ms/sample - loss

In [12]:
model_length.evaluate(x_test, y_test, verbose=1)

10440/1 [===============================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================

5.294550879430954

In [13]:
for i in range(1, 10):
    x_val[i]
    y_val[i]
    testing_x = np.array(x_val[i]).reshape(-1, x_val.shape[1], 1)
    testing_y = np.array(y_val[i]).reshape(y_val.shape[1],1)
    x = model_length.predict(testing_x).round(1)
    print(f"Predicted {x}, Expected : {testing_y}")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[76.4]], Expected : [[75]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[131.9]], Expected : [[140]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[86.7]], Expected : [[85]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[94.]], Expected : [[95]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[55.8]], Expected : [[45]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[42.3]], Expected : [[40]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[116.8]], Expected : [[130]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[114.9]], Expected : [[130]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[105.4]], Expected : [[105]]


<IPython.core.display.Javascript object>

In [1]:
df = pd.read_csv("data_one_per_bit_largest_full_jump10.csv")

<IPython.core.display.Javascript object>

In [2]:
df.drop(['Type_of_modulation'], axis=1, inplace = True)

In [3]:
x_train, x_test, y_train, y_test = train_test_split(df.drop(['SMF_length'], axis = 1),
                                                df['SMF_length'],
                                                test_size = 0.3,
                                                random_state = 101)

<IPython.core.display.Javascript object>

In [4]:
X_train, x_val, Y_train, y_val = train_test_split(df.drop(['SMF_length'], axis = 1),
                                                df['SMF_length'],
                                                test_size = 0.3,
                                                random_state = 101)

<IPython.core.display.Javascript object>

In [5]:
x_train = np.array(x_train).reshape(-1, x_train.shape[1], 1)
x_test = np.array(x_test).reshape(-1,  x_test.shape[1], 1)
y_train = np.array(y_train).reshape(y_train.shape[0], 1)
y_test = np.array(y_test).reshape(y_test.shape[0], 1)
x_val = np.array(x_val).reshape(-1,  x_val.shape[1], 1)
y_val = np.array(y_val).reshape(y_val.shape[0], 1)
x_train.shape

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

(12600, 2000, 1)

In [6]:
model_length_jump10 = tf.keras.Sequential([
        tf.keras.layers.Conv1D(32,kernel_size = 3, activation = 'relu', input_shape = x_train.shape[1:]),
        tf.keras.layers.MaxPooling1D(pool_size=3),
        tf.keras.layers.Dropout(0.2),
        tf.keras.layers.Conv1D(64,kernel_size = 3, activation = 'relu'),
        tf.keras.layers.MaxPooling1D(pool_size=3),
        tf.keras.layers.Dropout(0.2),
        tf.keras.layers.Conv1D(64,kernel_size = 3, activation = 'relu'),
        tf.keras.layers.MaxPooling1D(pool_size=3),
        tf.keras.layers.Dropout(0.2),
        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(1024, activation = 'relu'),
        tf.keras.layers.Dense(512, activation = 'relu'),
        tf.keras.layers.Dense(1, activation = 'relu')
        ])

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [7]:
model_length_jump10.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv1d (Conv1D)              (None, 1998, 32)          128       
_________________________________________________________________
max_pooling1d (MaxPooling1D) (None, 666, 32)           0         
_________________________________________________________________
dropout (Dropout)            (None, 666, 32)           0         
_________________________________________________________________
conv1d_1 (Conv1D)            (None, 664, 64)           6208      
_________________________________________________________________
max_pooling1d_1 (MaxPooling1 (None, 221, 64)           0         
_________________________________________________________________
dropout_1 (Dropout)          (None, 221, 64)           0         
_________________________________________________________________
conv1d_2 (Conv1D)            (None, 219, 64)           1

In [8]:
model_length_jump10.compile(loss='mae', optimizer='adam')

In [9]:
# model_length_jump10.fit(x_train, y_train, batch_size=64, verbose=1, epochs=50,validation_data=(x_val, y_val), callbacks = [earlystopping])
model_length_jump10.fit(x_train, y_train, batch_size=64, verbose=1, epochs=50,validation_data=(x_val, y_val))

Train on 12600 samples, validate on 5400 samples
Epoch 1/50
12600/12600 [==============================] - 43s 3ms/sample - loss: 25.4763 - val_loss: 20.9383
Epoch 2/50
12600/12600 [==============================] - 40s 3ms/sample - loss: 16.4879 - val_loss: 15.3688
Epoch 3/50
12600/12600 [==============================] - 42s 3ms/sample - loss: 14.7962 - val_loss: 12.8996
Epoch 4/50
12600/12600 [==============================] - 41s 3ms/sample - loss: 11.6874 - val_loss: 11.2659
Epoch 5/50
12600/12600 [==============================] - 41s 3ms/sample - loss: 11.1319 - val_loss: 11.3032
Epoch 6/50
12600/12600 [==============================] - 40s 3ms/sample - loss: 9.4394 - val_loss: 8.5427
Epoch 7/50
12600/12600 [==============================] - 40s 3ms/sample - loss: 9.2407 - val_loss: 9.5370
Epoch 8/50
12600/12600 [==============================] - 42s 3ms/sample - loss: 9.2930 - val_loss: 11.5240
Epoch 9/50
12600/12600 [==============================] - 44s 3ms/sample - loss: 9.0

In [10]:
model_length_jump10.evaluate(x_test, y_test, verbose=1)

5400/1 [================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================

6.5521555151762785

In [11]:
for i in range(1, 10):
    x_val[i]
    y_val[i]
    testing_x = np.array(x_val[i]).reshape(-1, x_val.shape[1], 1)
    testing_y = np.array(y_val[i]).reshape(y_val.shape[1],1)
    x = model_length_jump10.predict(testing_x).round(1)
    print(f"Predicted {x}, Expected : {testing_y}")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[52.6]], Expected : [[30]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[13.7]], Expected : [[10]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[124.7]], Expected : [[140]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[31.]], Expected : [[30]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[28.7]], Expected : [[10]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[132.6]], Expected : [[130]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[72.1]], Expected : [[80]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[38.3]], Expected : [[40]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[73.7]], Expected : [[80]]


In [11]:
df = pd.read_csv("data_one_per_bit_largest_full.csv")

<IPython.core.display.Javascript object>

In [12]:
df.drop(['Type_of_modulation'], axis=1, inplace = True)

In [13]:
x_train, x_test, y_train, y_test = train_test_split(df.drop(['SMF_length'], axis = 1),
                                                df['SMF_length'],
                                                test_size = 0.3,
                                                random_state = 101)

<IPython.core.display.Javascript object>

In [14]:
X_train, x_val, Y_train, y_val = train_test_split(df.drop(['SMF_length'], axis = 1),
                                                df['SMF_length'],
                                                test_size = 0.3,
                                                random_state = 101)

<IPython.core.display.Javascript object>

In [5]:
from sklearn.preprocessing import StandardScaler
ss = StandardScaler()
x_train = ss.fit_transform(x_train)
x_test = ss.fit_transform(x_test)
x_val = ss.fit_transform(x_val)

In [15]:
x_train = np.array(x_train).reshape(-1, x_train.shape[1], 1)
x_test = np.array(x_test).reshape(-1,  x_test.shape[1], 1)
y_train = np.array(y_train).reshape(y_train.shape[0], 1)
y_test = np.array(y_test).reshape(y_test.shape[0], 1)
x_val = np.array(x_val).reshape(-1,  x_val.shape[1], 1)
y_val = np.array(y_val).reshape(y_val.shape[0], 1)
x_train.shape

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

(24360, 2000, 1)

In [7]:
model_length_jump10r = tf.keras.Sequential([
        tf.keras.layers.Conv1D(32,kernel_size = 3, activation = 'relu', input_shape = x_train.shape[1:]),
        tf.keras.layers.MaxPooling1D(pool_size=3),
        tf.keras.layers.Dropout(0.3),
        tf.keras.layers.Conv1D(128,kernel_size = 3, activation = 'relu'),
        tf.keras.layers.MaxPooling1D(pool_size=3),
        tf.keras.layers.Dropout(0.3),
        tf.keras.layers.Conv1D(256,kernel_size = 3, activation = 'relu'),
        tf.keras.layers.MaxPooling1D(pool_size=3),
        tf.keras.layers.Dropout(0.2),
        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(1024, activation = 'relu'),
        tf.keras.layers.Dense(512, activation = 'relu'),
        tf.keras.layers.Dense(1, activation = 'relu')
        ])

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [8]:
# # Either run this or the one above... see which one gets better results

# model_length_jump10r = tf.keras.Sequential([
#         tf.keras.layers.Dense(1024, activation = 'relu', input_shape = x_train.shape[1:]),
#         tf.keras.layers.Dense(1024, activation = 'relu'),
#         tf.keras.layers.Dense(1, activation = 'relu')
#         ])

In [9]:
model_length_jump10r.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv1d (Conv1D)              (None, 1998, 32)          128       
_________________________________________________________________
max_pooling1d (MaxPooling1D) (None, 666, 32)           0         
_________________________________________________________________
dropout (Dropout)            (None, 666, 32)           0         
_________________________________________________________________
conv1d_1 (Conv1D)            (None, 664, 128)          12416     
_________________________________________________________________
max_pooling1d_1 (MaxPooling1 (None, 221, 128)          0         
_________________________________________________________________
dropout_1 (Dropout)          (None, 221, 128)          0         
_________________________________________________________________
conv1d_2 (Conv1D)            (None, 219, 256)          9

In [10]:
model_length_jump10r.compile(loss='mae', optimizer='adam')

In [11]:
# model_length_jump10.fit(x_train, y_train, batch_size=64, verbose=1, epochs=50,validation_data=(x_val, y_val), callbacks = [earlystopping])
model_length_jump10r.fit(x_train, y_train, batch_size=32, verbose=1, epochs=25,validation_data=(x_val, y_val))

Train on 24360 samples, validate on 10440 samples
Epoch 1/25
24360/24360 [==============================] - 219s 9ms/sample - loss: 14.3781 - val_loss: 9.3970
Epoch 2/25
24360/24360 [==============================] - 226s 9ms/sample - loss: 7.8159 - val_loss: 8.4880
Epoch 3/25
24360/24360 [==============================] - 231s 9ms/sample - loss: 7.1855 - val_loss: 6.9134
Epoch 4/25
24360/24360 [==============================] - 231s 10ms/sample - loss: 6.5643 - val_loss: 7.8693
Epoch 5/25
24360/24360 [==============================] - 232s 10ms/sample - loss: 6.1409 - val_loss: 8.6418
Epoch 6/25
24360/24360 [==============================] - 230s 9ms/sample - loss: 5.8505 - val_loss: 6.3810
Epoch 7/25
24360/24360 [==============================] - 228s 9ms/sample - loss: 5.5309 - val_loss: 7.0384
Epoch 8/25
24360/24360 [==============================] - 231s 9ms/sample - loss: 5.4895 - val_loss: 5.2151
Epoch 9/25
24360/24360 [==============================] - 230s 9ms/sample - loss: 5

In [12]:
model_length_jump10r.evaluate(x_test, y_test, verbose=1)

10440/1 [===============================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================

4.913290165575985

In [13]:
for i in range(1, 20):
    x_val[i]
    y_val[i]
    testing_x = np.array(x_val[i]).reshape(-1, x_val.shape[1], 1)
    testing_y = np.array(y_val[i]).reshape(y_val.shape[1],1)
    x = model_length_jump10r.predict(testing_x).round(1)
    print(f"Predicted {x}, Expected : {testing_y}")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[77.6]], Expected : [[75]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[135.3]], Expected : [[140]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[88.]], Expected : [[85]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[99.1]], Expected : [[95]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[55.8]], Expected : [[45]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[42.7]], Expected : [[40]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[120.9]], Expected : [[130]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[117.]], Expected : [[130]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[108.8]], Expected : [[105]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[97.5]], Expected : [[95]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[135.4]], Expected : [[135]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[83.6]], Expected : [[80]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[98.4]], Expected : [[95]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[141.3]], Expected : [[145]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[124.7]], Expected : [[125]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[21.7]], Expected : [[25]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[52.2]], Expected : [[50]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[22.5]], Expected : [[20]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[49.4]], Expected : [[50]]


In [14]:
model_length_jump1 = tf.keras.Sequential([
        tf.keras.layers.Conv1D(64,kernel_size = 3, activation = 'relu', input_shape = x_train.shape[1:]),
        tf.keras.layers.MaxPooling1D(pool_size=3),
        tf.keras.layers.Dropout(0.3),
        tf.keras.layers.Conv1D(256,kernel_size = 3, activation = 'relu'),
        tf.keras.layers.MaxPooling1D(pool_size=3),
        tf.keras.layers.Dropout(0.3),
        tf.keras.layers.Conv1D(256,kernel_size = 3, activation = 'relu'),
        tf.keras.layers.MaxPooling1D(pool_size=3),
        tf.keras.layers.Dropout(0.2),
        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(1024, activation = 'relu'),
        tf.keras.layers.Dense(512, activation = 'relu'),
        tf.keras.layers.Dense(256, activation = 'relu'),
        tf.keras.layers.Dense(1, activation = 'relu')
        ])

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [15]:
model_length_jump1.compile(loss='mae', optimizer='adam')

In [16]:
model_length_jump1.fit(x_train, y_train, batch_size=32, verbose=1, epochs=25,validation_data=(x_val, y_val))

Train on 24360 samples, validate on 10440 samples
Epoch 1/25
24360/24360 [==============================] - 321s 13ms/sample - loss: 13.0335 - val_loss: 8.5250
Epoch 2/25
24360/24360 [==============================] - 316s 13ms/sample - loss: 7.8264 - val_loss: 10.5246
Epoch 3/25
24360/24360 [==============================] - 322s 13ms/sample - loss: 7.0610 - val_loss: 6.1210
Epoch 4/25
24360/24360 [==============================] - 331s 14ms/sample - loss: 6.4781 - val_loss: 6.0019
Epoch 5/25
24360/24360 [==============================] - 330s 14ms/sample - loss: 6.1453 - val_loss: 5.6632
Epoch 6/25
24360/24360 [==============================] - 330s 14ms/sample - loss: 5.8414 - val_loss: 6.1750
Epoch 7/25
24360/24360 [==============================] - 332s 14ms/sample - loss: 5.6195 - val_loss: 5.0530
Epoch 8/25
24360/24360 [==============================] - 330s 14ms/sample - loss: 5.4462 - val_loss: 4.5820
Epoch 9/25
24360/24360 [==============================] - 333s 14ms/sample -

In [18]:
model_length_jump1.evaluate(x_test, y_test, verbose=1)


10440/1 [===============================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================

3.6666662121184483

In [19]:
for i in range(1, 20):
    x_val[i]
    y_val[i]
    testing_x = np.array(x_val[i]).reshape(-1, x_val.shape[1], 1)
    testing_y = np.array(y_val[i]).reshape(y_val.shape[1],1)
    x = model_length_jump1.predict(testing_x).round(1)
    print(f"Predicted {x}, Expected : {testing_y}")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[73.]], Expected : [[75]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[137.4]], Expected : [[140]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[82.6]], Expected : [[85]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[96.6]], Expected : [[95]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[50.8]], Expected : [[45]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[37.7]], Expected : [[40]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[122.3]], Expected : [[130]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[120.8]], Expected : [[130]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[105.8]], Expected : [[105]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[92.8]], Expected : [[95]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[135.4]], Expected : [[135]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[77.1]], Expected : [[80]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[98.6]], Expected : [[95]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[144.3]], Expected : [[145]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[124.8]], Expected : [[125]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[22.9]], Expected : [[25]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[48.6]], Expected : [[50]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[15.9]], Expected : [[20]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[51.9]], Expected : [[50]]


In [20]:
model_length_jump2 = tf.keras.Sequential([
        tf.keras.layers.Conv1D(64,kernel_size = 3, activation = 'relu', input_shape = x_train.shape[1:]),
        tf.keras.layers.MaxPooling1D(pool_size=3),
        tf.keras.layers.Dropout(0.3),
        tf.keras.layers.Conv1D(128,kernel_size = 3, activation = 'relu'),
        tf.keras.layers.MaxPooling1D(pool_size=3),
        tf.keras.layers.Dropout(0.3),
        tf.keras.layers.Conv1D(256,kernel_size = 3, activation = 'relu'),
        tf.keras.layers.MaxPooling1D(pool_size=3),
        tf.keras.layers.Dropout(0.3),
        tf.keras.layers.Conv1D(256,kernel_size = 3, activation = 'relu'),
        tf.keras.layers.MaxPooling1D(pool_size=3),
        tf.keras.layers.Dropout(0.2),
        tf.keras.layers.Conv1D(64,kernel_size = 3, activation = 'relu'),
        tf.keras.layers.MaxPooling1D(pool_size=3),
        tf.keras.layers.Dropout(0.3),
        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(1024, activation = 'relu'),
        tf.keras.layers.Dense(512, activation = 'relu'),
        tf.keras.layers.Dense(256, activation = 'relu'),
        tf.keras.layers.Dense(128, activation = 'relu'),
        tf.keras.layers.Dense(1, activation = 'relu')
        ])

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [21]:
model_length_jump2.compile(loss='mae', optimizer='adam')

In [22]:
model_length_jump2.summary()

Model: "sequential_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv1d_6 (Conv1D)            (None, 1998, 64)          256       
_________________________________________________________________
max_pooling1d_6 (MaxPooling1 (None, 666, 64)           0         
_________________________________________________________________
dropout_6 (Dropout)          (None, 666, 64)           0         
_________________________________________________________________
conv1d_7 (Conv1D)            (None, 664, 128)          24704     
_________________________________________________________________
max_pooling1d_7 (MaxPooling1 (None, 221, 128)          0         
_________________________________________________________________
dropout_7 (Dropout)          (None, 221, 128)          0         
_________________________________________________________________
conv1d_8 (Conv1D)            (None, 219, 256)         

In [23]:
model_length_jump2.fit(x_train, y_train, batch_size=32, verbose=1, epochs=50,validation_data=(x_val, y_val))

Train on 24360 samples, validate on 10440 samples
Epoch 1/50
24360/24360 [==============================] - 194s 8ms/sample - loss: 16.1387 - val_loss: 12.4075
Epoch 2/50
24360/24360 [==============================] - 188s 8ms/sample - loss: 9.5510 - val_loss: 9.9934
Epoch 3/50
24360/24360 [==============================] - 193s 8ms/sample - loss: 8.5167 - val_loss: 11.2932
Epoch 4/50
24360/24360 [==============================] - 195s 8ms/sample - loss: 7.9164 - val_loss: 8.5362
Epoch 5/50
24360/24360 [==============================] - 193s 8ms/sample - loss: 7.4292 - val_loss: 11.9437
Epoch 6/50
24360/24360 [==============================] - 194s 8ms/sample - loss: 7.0959 - val_loss: 9.9536
Epoch 7/50
24360/24360 [==============================] - 192s 8ms/sample - loss: 6.8528 - val_loss: 10.0949
Epoch 8/50
24360/24360 [==============================] - 188s 8ms/sample - loss: 6.5739 - val_loss: 12.0250
Epoch 9/50
24360/24360 [==============================] - 190s 8ms/sample - loss

In [24]:
model_length_jump2.evaluate(x_test, y_test, verbose=1)

10440/1 [===============================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================

7.794948550111033

In [29]:
for i in range(1, 20):
    x_val[i]
    y_val[i]
    testing_x = np.array(x_val[i]).reshape(-1, x_val.shape[1], 1)
    testing_y = np.array(y_val[i]).reshape(y_val.shape[1],1)
    x = model_length_jump2.predict(testing_x).round(1)
    print(f"Predicted {x}, Expected : {testing_y}")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[80.1]], Expected : [[75]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[138.9]], Expected : [[140]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[94.8]], Expected : [[85]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[95.5]], Expected : [[95]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[56.6]], Expected : [[45]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[60.6]], Expected : [[40]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[126.1]], Expected : [[130]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[120.7]], Expected : [[130]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[113.1]], Expected : [[105]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[106.6]], Expected : [[95]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[134.6]], Expected : [[135]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[91.8]], Expected : [[80]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[96.3]], Expected : [[95]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[147.2]], Expected : [[145]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[127.2]], Expected : [[125]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[46.5]], Expected : [[25]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[67.7]], Expected : [[50]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[40.1]], Expected : [[20]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[54.8]], Expected : [[50]]


In [26]:
model_length_jump1.fit(x_train, y_train, batch_size=32, verbose=1, epochs=50,validation_data=(x_val, y_val))

Train on 24360 samples, validate on 10440 samples
Epoch 1/50
24360/24360 [==============================] - 351s 14ms/sample - loss: 3.8868 - val_loss: 3.7069
Epoch 2/50
24360/24360 [==============================] - 371s 15ms/sample - loss: 3.6905 - val_loss: 4.4143
Epoch 3/50
24360/24360 [==============================] - 361s 15ms/sample - loss: 3.7901 - val_loss: 3.7802
Epoch 4/50
24360/24360 [==============================] - 361s 15ms/sample - loss: 3.7658 - val_loss: 4.5722
Epoch 5/50
24360/24360 [==============================] - 328s 13ms/sample - loss: 3.5593 - val_loss: 3.3665
Epoch 6/50
24360/24360 [==============================] - 397s 16ms/sample - loss: 3.5463 - val_loss: 3.2867
Epoch 7/50
24360/24360 [==============================] - 295s 12ms/sample - loss: 3.3954 - val_loss: 3.7198
Epoch 8/50
24360/24360 [==============================] - 310s 13ms/sample - loss: 3.5104 - val_loss: 4.9708
Epoch 9/50
24360/24360 [==============================] - 306s 13ms/sample - l

In [27]:
model_length_jump1.evaluate(x_test, y_test, verbose=1)

10440/1 [===============================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================

2.912765268347729

In [28]:
for i in range(1, 20):
    x_val[i]
    y_val[i]
    testing_x = np.array(x_val[i]).reshape(-1, x_val.shape[1], 1)
    testing_y = np.array(y_val[i]).reshape(y_val.shape[1],1)
    x = model_length_jump1.predict(testing_x).round(1)
    print(f"Predicted {x}, Expected : {testing_y}")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[74.7]], Expected : [[75]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[146.1]], Expected : [[140]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[84.5]], Expected : [[85]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[103.1]], Expected : [[95]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[51.7]], Expected : [[45]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[43.6]], Expected : [[40]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[126.9]], Expected : [[130]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[129.3]], Expected : [[130]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[104.7]], Expected : [[105]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[93.4]], Expected : [[95]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[135.7]], Expected : [[135]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[80.]], Expected : [[80]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[101.2]], Expected : [[95]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[143.]], Expected : [[145]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[131.3]], Expected : [[125]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[23.1]], Expected : [[25]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[52.1]], Expected : [[50]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[21.]], Expected : [[20]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[49.5]], Expected : [[50]]


In [30]:
model_length_jump1.fit(x_train, y_train, batch_size=32, verbose=1, epochs=25,validation_data=(x_val, y_val))

Train on 24360 samples, validate on 10440 samples
Epoch 1/25
24360/24360 [==============================] - 319s 13ms/sample - loss: 2.3954 - val_loss: 3.2303
Epoch 2/25
24360/24360 [==============================] - 312s 13ms/sample - loss: 2.4532 - val_loss: 3.1163
Epoch 3/25
24360/24360 [==============================] - 308s 13ms/sample - loss: 2.4499 - val_loss: 3.5328
Epoch 4/25
24360/24360 [==============================] - 307s 13ms/sample - loss: 2.4099 - val_loss: 4.0143
Epoch 5/25
24360/24360 [==============================] - 305s 13ms/sample - loss: 2.4132 - val_loss: 2.9678
Epoch 6/25
24360/24360 [==============================] - 305s 13ms/sample - loss: 2.4069 - val_loss: 2.8869
Epoch 7/25
24360/24360 [==============================] - 289s 12ms/sample - loss: 2.4120 - val_loss: 3.0746
Epoch 8/25
24360/24360 [==============================] - 297s 12ms/sample - loss: 2.3160 - val_loss: 3.0392
Epoch 9/25
24360/24360 [==============================] - 290s 12ms/sample - l

In [7]:
model_length_jump3 = tf.keras.Sequential([
        tf.keras.layers.Conv1D(32,kernel_size = 3, activation = 'relu', input_shape = x_train.shape[1:]),
        tf.keras.layers.MaxPooling1D(pool_size=3),
        tf.keras.layers.Dropout(0.3),
        tf.keras.layers.Conv1D(64,kernel_size = 3, activation = 'relu'),
        tf.keras.layers.MaxPooling1D(pool_size=3),
        tf.keras.layers.Dropout(0.3),
        tf.keras.layers.Conv1D(128,kernel_size = 3, activation = 'relu'),
        tf.keras.layers.MaxPooling1D(pool_size=3),
        tf.keras.layers.Dropout(0.3),
        tf.keras.layers.Conv1D(256,kernel_size = 3, activation = 'relu'),
        tf.keras.layers.MaxPooling1D(pool_size=3),
        tf.keras.layers.Dropout(0.2),
        tf.keras.layers.Conv1D(512,kernel_size = 3, activation = 'relu'),
        tf.keras.layers.MaxPooling1D(pool_size=3),
        tf.keras.layers.Dropout(0.3),
        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(1024, activation = 'relu'),
        tf.keras.layers.Dense(512, activation = 'relu'),
        tf.keras.layers.Dense(1, activation = 'relu')
        ])

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [8]:
model_length_jump3.compile(loss='mae', optimizer='adam')

In [9]:
model_length_jump3.fit(x_train, y_train, batch_size=32, verbose=1, epochs=10,validation_data=(x_val, y_val))

Train on 24360 samples, validate on 10440 samples
Epoch 1/10
24360/24360 [==============================] - 173s 7ms/sample - loss: 14.2879 - val_loss: 17.8206
Epoch 2/10
24360/24360 [==============================] - 167s 7ms/sample - loss: 8.5059 - val_loss: 13.3595
Epoch 3/10
24360/24360 [==============================] - 166s 7ms/sample - loss: 7.8032 - val_loss: 18.0787
Epoch 4/10
24360/24360 [==============================] - 161s 7ms/sample - loss: 7.3356 - val_loss: 16.4004
Epoch 5/10
24360/24360 [==============================] - 164s 7ms/sample - loss: 6.8585 - val_loss: 10.8889
Epoch 6/10
24360/24360 [==============================] - 169s 7ms/sample - loss: 6.6826 - val_loss: 12.9103
Epoch 7/10
24360/24360 [==============================] - 168s 7ms/sample - loss: 6.3107 - val_loss: 10.9374
Epoch 8/10
24360/24360 [==============================] - 168s 7ms/sample - loss: 6.1391 - val_loss: 12.2943
Epoch 9/10
24360/24360 [==============================] - 167s 7ms/sample - l

In [13]:
model_length_jump4 = tf.keras.Sequential([
        tf.keras.layers.Conv1D(512,kernel_size = 3, activation = 'relu', input_shape = x_train.shape[1:]),
        tf.keras.layers.MaxPooling1D(pool_size=3),
        tf.keras.layers.Dropout(0.3),
        tf.keras.layers.Conv1D(256,kernel_size = 3, activation = 'relu'),
        tf.keras.layers.MaxPooling1D(pool_size=3),
        tf.keras.layers.Dropout(0.3),
        tf.keras.layers.Conv1D(128,kernel_size = 3, activation = 'relu'),
        tf.keras.layers.MaxPooling1D(pool_size=3),
        tf.keras.layers.Dropout(0.3),
        tf.keras.layers.Conv1D(128,kernel_size = 3, activation = 'relu'),
        tf.keras.layers.MaxPooling1D(pool_size=3),
        tf.keras.layers.Dropout(0.2),
        tf.keras.layers.Conv1D(64,kernel_size = 3, activation = 'relu'),
        tf.keras.layers.MaxPooling1D(pool_size=3),
        tf.keras.layers.Dropout(0.3),
        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(1024, activation = 'relu'),
        tf.keras.layers.Dense(512, activation = 'relu'),
        tf.keras.layers.Dense(256, activation = 'relu'),
        tf.keras.layers.Dense(1, activation = 'relu')
        ])

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [14]:
model_length_jump4.compile(loss='mae', optimizer='adam')

In [15]:
model_length_jump4.fit(x_train, y_train, batch_size=32, verbose=1, epochs=3,validation_data=(x_val, y_val))

Train on 24360 samples, validate on 10440 samples
Epoch 1/3
24360/24360 [==============================] - 678s 28ms/sample - loss: 15.4004 - val_loss: 20.3939
Epoch 2/3
24360/24360 [==============================] - 679s 28ms/sample - loss: 9.9198 - val_loss: 11.5264
Epoch 3/3
24360/24360 [==============================] - 695s 29ms/sample - loss: 8.9284 - val_loss: 12.7265


In [130]:
model_length_jump5 = tf.keras.Sequential([
        tf.keras.layers.Conv1D(64,kernel_size = 5, activation = 'relu', input_shape = x_train.shape[1:]),
        tf.keras.layers.Dropout(0.3),
        tf.keras.layers.Conv1D(32,kernel_size = 5, activation = 'relu'),
        tf.keras.layers.MaxPooling1D(pool_size=2),
        tf.keras.layers.Dropout(0.3),
        tf.keras.layers.Conv1D(16,kernel_size = 3, activation = 'relu'),
        tf.keras.layers.MaxPooling1D(pool_size=2),
        tf.keras.layers.Dropout(0.3),
        tf.keras.layers.Conv1D(16,kernel_size = 3, activation = 'relu'),         
        tf.keras.layers.MaxPooling1D(pool_size=2),
        tf.keras.layers.Dropout(0.3),
        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(64, activation = 'relu'),
        tf.keras.layers.Dense(32, activation = 'relu'),
        tf.keras.layers.Dense(16, activation = 'relu'),
        tf.keras.layers.Dense(1, activation = 'relu')
        ])

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [131]:
model_length_jump5.summary()

Model: "sequential_35"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv1d_187 (Conv1D)          (None, 1996, 64)          384       
_________________________________________________________________
dropout_187 (Dropout)        (None, 1996, 64)          0         
_________________________________________________________________
conv1d_188 (Conv1D)          (None, 1992, 32)          10272     
_________________________________________________________________
max_pooling1d_142 (MaxPoolin (None, 996, 32)           0         
_________________________________________________________________
dropout_188 (Dropout)        (None, 996, 32)           0         
_________________________________________________________________
conv1d_189 (Conv1D)          (None, 992, 16)           2576      
_________________________________________________________________
max_pooling1d_143 (MaxPoolin (None, 496, 16)         

In [132]:
model_length_jump5.compile(loss='mae', optimizer='adam')

In [133]:
model_length_jump5.fit(x_train, y_train, batch_size=32, verbose=1, epochs=100,validation_data=(x_val, y_val))

Train on 24360 samples, validate on 10440 samples
Epoch 1/100
24360/24360 [==============================] - 227s 9ms/sample - loss: 15.9513 - val_loss: 20.9786
Epoch 2/100
24360/24360 [==============================] - 227s 9ms/sample - loss: 8.4381 - val_loss: 14.7710
Epoch 3/100
24360/24360 [==============================] - 227s 9ms/sample - loss: 7.7265 - val_loss: 15.0943
Epoch 4/100
24360/24360 [==============================] - 225s 9ms/sample - loss: 7.2183 - val_loss: 16.4337
Epoch 5/100
24360/24360 [==============================] - 226s 9ms/sample - loss: 6.9909 - val_loss: 17.7638
Epoch 6/100
24360/24360 [==============================] - 224s 9ms/sample - loss: 6.8036 - val_loss: 7.5635
Epoch 7/100
24360/24360 [==============================] - 228s 9ms/sample - loss: 6.5322 - val_loss: 9.4805
Epoch 8/100
24360/24360 [==============================] - 234s 10ms/sample - loss: 6.2109 - val_loss: 10.2205
Epoch 9/100
24360/24360 [==============================] - 226s 9ms/sa

Epoch 75/100
24360/24360 [==============================] - 227s 9ms/sample - loss: 3.8314 - val_loss: 9.4092
Epoch 76/100
24360/24360 [==============================] - 224s 9ms/sample - loss: 3.8437 - val_loss: 8.2768
Epoch 77/100
24360/24360 [==============================] - 227s 9ms/sample - loss: 3.7918 - val_loss: 8.1574
Epoch 78/100
24360/24360 [==============================] - 226s 9ms/sample - loss: 3.8162 - val_loss: 8.7743
Epoch 79/100
24360/24360 [==============================] - 227s 9ms/sample - loss: 3.7412 - val_loss: 6.9805
Epoch 80/100
24360/24360 [==============================] - 224s 9ms/sample - loss: 3.7976 - val_loss: 7.8419
Epoch 81/100
24360/24360 [==============================] - 226s 9ms/sample - loss: 3.7919 - val_loss: 7.3447
Epoch 82/100
24360/24360 [==============================] - 224s 9ms/sample - loss: 3.7274 - val_loss: 9.2012
Epoch 83/100
24360/24360 [==============================] - 225s 9ms/sample - loss: 3.7114 - val_loss: 7.6515
Epoch 84/1

In [134]:
model_length_jump5.evaluate(x_test, y_test, verbose=1)

10440/1 [===============================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================

8.260895342114328

In [135]:
for i in range(1, 20):
    x_val[i]
    y_val[i]
    testing_x = np.array(x_val[i]).reshape(-1, x_val.shape[1], 1)
    testing_y = np.array(y_val[i]).reshape(y_val.shape[1],1)
    x = model_length_jump5.predict(testing_x).round(1)
    print(f"Predicted {x}, Expected : {testing_y}")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[69.8]], Expected : [[75]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[126.5]], Expected : [[140]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[75.3]], Expected : [[85]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[86.]], Expected : [[95]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[35.9]], Expected : [[45]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[38.7]], Expected : [[40]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[121.8]], Expected : [[130]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[109.2]], Expected : [[130]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[95.4]], Expected : [[105]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[88.5]], Expected : [[95]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[118.]], Expected : [[135]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[72.6]], Expected : [[80]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[83.8]], Expected : [[95]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[133.]], Expected : [[145]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[120.7]], Expected : [[125]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[22.2]], Expected : [[25]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[47.4]], Expected : [[50]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[17.]], Expected : [[20]]


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Predicted [[47.1]], Expected : [[50]]


In [136]:
model_length_jump6 = tf.keras.Sequential([
        tf.keras.layers.Conv1D(64,kernel_size = 3, activation = 'relu', input_shape = x_train.shape[1:]),
        tf.keras.layers.Dropout(0.3),
        tf.keras.layers.Conv1D(32,kernel_size = 3, activation = 'relu'),
        tf.keras.layers.MaxPooling1D(pool_size=2),
        tf.keras.layers.Dropout(0.3),
        tf.keras.layers.Conv1D(16,kernel_size = 3, activation = 'relu'),
        tf.keras.layers.MaxPooling1D(pool_size=2),
        tf.keras.layers.Dropout(0.3),
        tf.keras.layers.Conv1D(16,kernel_size = 3, activation = 'relu'),         
        tf.keras.layers.MaxPooling1D(pool_size=2),
        tf.keras.layers.Dropout(0.3),
        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(64, activation = 'relu'),
        tf.keras.layers.Dense(32, activation = 'relu'),
        tf.keras.layers.Dense(16, activation = 'relu'),
        tf.keras.layers.Dense(1, activation = 'relu')
        ])

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [137]:
model_length_jump6.compile(loss='mae', optimizer='adam')

In [138]:
model_length_jump6.fit(x_train, y_train, batch_size=32, verbose=1, epochs=100,validation_data=(x_val, y_val))

Train on 24360 samples, validate on 10440 samples
Epoch 1/100
24360/24360 [==============================] - 146s 6ms/sample - loss: 15.7148 - val_loss: 18.2036
Epoch 2/100
24360/24360 [==============================] - 135s 6ms/sample - loss: 8.8868 - val_loss: 19.1409
Epoch 3/100
24360/24360 [==============================] - 128s 5ms/sample - loss: 8.0527 - val_loss: 19.5651
Epoch 4/100
24360/24360 [==============================] - 129s 5ms/sample - loss: 7.6797 - val_loss: 13.1060
Epoch 5/100
24360/24360 [==============================] - 132s 5ms/sample - loss: 7.4138 - val_loss: 17.1707
Epoch 6/100
24360/24360 [==============================] - 171s 7ms/sample - loss: 6.9797 - val_loss: 8.7257
Epoch 7/100
24360/24360 [==============================] - 167s 7ms/sample - loss: 6.6420 - val_loss: 13.8138
Epoch 8/100
24360/24360 [==============================] - 177s 7ms/sample - loss: 6.4286 - val_loss: 9.2451
Epoch 9/100
24360/24360 [==============================] - 169s 7ms/sam

Epoch 75/100
24360/24360 [==============================] - 195s 8ms/sample - loss: 4.0269 - val_loss: 6.3753
Epoch 76/100
24360/24360 [==============================] - 196s 8ms/sample - loss: 4.0526 - val_loss: 5.7319
Epoch 77/100
24360/24360 [==============================] - 195s 8ms/sample - loss: 4.0412 - val_loss: 5.7212
Epoch 78/100
24360/24360 [==============================] - 195s 8ms/sample - loss: 4.0474 - val_loss: 5.2836
Epoch 79/100
24360/24360 [==============================] - 195s 8ms/sample - loss: 4.0529 - val_loss: 5.4833
Epoch 80/100
24360/24360 [==============================] - 195s 8ms/sample - loss: 3.9733 - val_loss: 6.0251
Epoch 81/100
24360/24360 [==============================] - 195s 8ms/sample - loss: 4.0559 - val_loss: 6.2796
Epoch 82/100
24360/24360 [==============================] - 195s 8ms/sample - loss: 4.0353 - val_loss: 6.1676
Epoch 83/100
24360/24360 [==============================] - 195s 8ms/sample - loss: 4.0082 - val_loss: 5.1461
Epoch 84/1

In [139]:
model_length_jump6.evaluate(x_test, y_test, verbose=1)

10440/1 [===============================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================

5.832125606938797

In [140]:
model_length_jump6s = tf.keras.Sequential([
        tf.keras.layers.Conv1D(64,kernel_size = 7, activation = 'relu', input_shape = x_train.shape[1:]),
        tf.keras.layers.Dropout(0.3),
        tf.keras.layers.Conv1D(32,kernel_size = 7, activation = 'relu'),
        tf.keras.layers.MaxPooling1D(pool_size=2),
        tf.keras.layers.Dropout(0.3),
        tf.keras.layers.Conv1D(16,kernel_size = 3, activation = 'relu'),
        tf.keras.layers.MaxPooling1D(pool_size=2),
        tf.keras.layers.Dropout(0.3),
        tf.keras.layers.Conv1D(16,kernel_size = 3, activation = 'relu'),         
        tf.keras.layers.MaxPooling1D(pool_size=2),
        tf.keras.layers.Dropout(0.3),
        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(64, activation = 'relu'),
        tf.keras.layers.Dense(32, activation = 'relu'),
        tf.keras.layers.Dense(16, activation = 'relu'),
        tf.keras.layers.Dense(1, activation = 'relu')
        ])

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [141]:
model_length_jump6s.compile(loss='mae', optimizer='adam')

In [143]:
model_length_jump6s.fit(x_train, y_train, batch_size=32, verbose=1, epochs=25,validation_data=(x_val, y_val))

Train on 24360 samples, validate on 10440 samples
Epoch 1/25
24360/24360 [==============================] - 155s 6ms/sample - loss: 6.9805 - val_loss: 13.6308
Epoch 2/25
24360/24360 [==============================] - 154s 6ms/sample - loss: 6.4638 - val_loss: 10.5501
Epoch 3/25
24360/24360 [==============================] - 156s 6ms/sample - loss: 6.2527 - val_loss: 12.5531
Epoch 4/25
24360/24360 [==============================] - 157s 6ms/sample - loss: 6.0118 - val_loss: 14.4207
Epoch 5/25
24360/24360 [==============================] - 153s 6ms/sample - loss: 5.9110 - val_loss: 12.0689
Epoch 6/25
24360/24360 [==============================] - 156s 6ms/sample - loss: 5.7676 - val_loss: 12.9131
Epoch 7/25
24360/24360 [==============================] - 159s 7ms/sample - loss: 5.7658 - val_loss: 10.6594
Epoch 8/25
24360/24360 [==============================] - 161s 7ms/sample - loss: 5.5586 - val_loss: 11.2563
Epoch 9/25
24360/24360 [==============================] - 163s 7ms/sample - lo

In [68]:
model_length_jump7 = tf.keras.Sequential([
        tf.keras.layers.Conv1D(16,kernel_size = 3, activation = 'relu', input_shape = x_train.shape[1:]),
        tf.keras.layers.MaxPooling1D(pool_size=3),
        tf.keras.layers.Dropout(0.3),
        tf.keras.layers.Conv1D(16,kernel_size = 3, activation = 'relu'),
        tf.keras.layers.MaxPooling1D(pool_size=3),
        tf.keras.layers.Dropout(0.3),
        tf.keras.layers.Conv1D(32,kernel_size = 3, activation = 'relu'),
        tf.keras.layers.MaxPooling1D(pool_size=3),
        tf.keras.layers.Dropout(0.3),
        tf.keras.layers.Conv1D(32,kernel_size = 3, activation = 'relu'),
        tf.keras.layers.MaxPooling1D(pool_size=3),
        tf.keras.layers.Dropout(0.3),
        tf.keras.layers.Conv1D(64,kernel_size = 3, activation = 'relu'),
        tf.keras.layers.MaxPooling1D(pool_size=3),
        tf.keras.layers.Dropout(0.2),
        tf.keras.layers.Conv1D(128,kernel_size = 3, activation = 'relu'),
        tf.keras.layers.MaxPooling1D(pool_size=3),
        tf.keras.layers.Dropout(0.3),
#         tf.keras.layers.Conv1D(128,kernel_size = 3, activation = 'relu'),
#         tf.keras.layers.MaxPooling1D(pool_size=3),
#         tf.keras.layers.Dropout(0.3),
#         tf.keras.layers.Conv1D(256,kernel_size = 3, activation = 'relu'),
#         tf.keras.layers.MaxPooling1D(pool_size=3),
#         tf.keras.layers.Dropout(0.3),
        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(256, activation = 'relu'),


        tf.keras.layers.Dense(1, activation = 'relu')
        ])

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [69]:
model_length_jump7.compile(loss='mae', optimizer='RMSProp')

In [70]:
model_length_jump7.fit(x_train, y_train, batch_size=32, verbose=1, epochs=3,validation_data=(x_val, y_val))

Train on 24360 samples, validate on 10440 samples
Epoch 1/3
24360/24360 [==============================] - 49s 2ms/sample - loss: 21.6680 - val_loss: 30.9906
Epoch 2/3
24360/24360 [==============================] - 31s 1ms/sample - loss: 15.2321 - val_loss: 20.6637
Epoch 3/3
24360/24360 [==============================] - 29s 1ms/sample - loss: 14.5203 - val_loss: 17.4061


In [27]:
model_length_jump1i = tf.keras.Sequential([
        tf.keras.layers.Conv1D(128,kernel_size = 3, activation = 'relu', input_shape = x_train.shape[1:]),
#         tf.keras.layers.MaxPooling1D(pool_size=3),
        tf.keras.layers.Dropout(0.3),
#         tf.keras.layers.Conv1D(16,kernel_size = 3, activation = 'relu'),
# #         tf.keras.layers.MaxPooling1D(pool_size=2),
#         tf.keras.layers.Dropout(0.3),
#         tf.keras.layers.Conv1D(16,kernel_size = 3, activation = 'relu'),
# #         tf.keras.layers.MaxPooling1D(pool_size=3),
#         tf.keras.layers.Dropout(0.2),
#         tf.keras.layers.Conv1D(64,kernel_size = 3, activation = 'relu'),
# #         tf.keras.layers.MaxPooling1D(pool_size=3),
#         tf.keras.layers.Dropout(0.4),
#         tf.keras.layers.Conv1D(128,kernel_size = 4, activation = 'relu'),
#         tf.keras.layers.MaxPooling1D(pool_size=3),
#         tf.keras.layers.Dropout(0.4),
#         tf.keras.layers.Conv1D(128,kernel_size = 4, activation = 'relu'),
#         tf.keras.layers.MaxPooling1D(pool_size=3),
#         tf.keras.layers.Dropout(0.4),
#         tf.keras.layers.Conv1D(128,kernel_size = 4, activation = 'relu'),
#         tf.keras.layers.MaxPooling1D(pool_size=3),
#         tf.keras.layers.Dropout(0.4),
        tf.keras.layers.Conv1D(256,kernel_size = 5, activation = 'relu'),
#         tf.keras.layers.MaxPooling1D(pool_size=3),
        tf.keras.layers.Dropout(0.4),
#         tf.keras.layers.Conv1D(256,kernel_size = 5, activation = 'relu'),
# #         tf.keras.layers.MaxPooling1D(pool_size=3),
#         tf.keras.layers.Dropout(0.4),
        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(512, activation = 'relu'),
        tf.keras.layers.Dense(128, activation = 'relu'),
        tf.keras.layers.Dense(32, activation = 'relu'),
        tf.keras.layers.Dense(1, activation = 'relu')
        ])

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [28]:
model_length_jump1i.summary()

Model: "sequential_3"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv1d_14 (Conv1D)           (None, 1998, 128)         512       
_________________________________________________________________
dropout_16 (Dropout)         (None, 1998, 128)         0         
_________________________________________________________________
conv1d_15 (Conv1D)           (None, 1994, 256)         164096    
_________________________________________________________________
dropout_17 (Dropout)         (None, 1994, 256)         0         
_________________________________________________________________
flatten_3 (Flatten)          (None, 510464)            0         
_________________________________________________________________
dense_12 (Dense)             (None, 512)               261358080 
_________________________________________________________________
dense_13 (Dense)             (None, 128)              

In [29]:
model_length_jump1i.compile(loss='mae', optimizer='adam')

In [ ]:
model_length_jump1i.fit(x_train, y_train, batch_size=32, verbose=1, epochs=1,validation_data=(x_val, y_val))

Train on 24360 samples, validate on 10440 samples
  224/24360 [..............................] - ETA: 47:24 - loss: 83.0134

# Observations on the type of model

1. More number of dense layers - longer time
2. More neurons in dense layers (2000+ 2 layers) - Stagnation in loss around 80 - too much
3. Smaller starting conv layer - faster training with lesser loss
4. For dense layers, best to have two/three layers with decreasing number of neurons... Any more, and the speed would be affected. However, if ready to compromise with training time, more dense layers give out better results.
5. Increasing the number of Conv layers beyond a point doesn't help. The model becomes both slow in training and in converging to a lower level of loss.

Tests with jump5 model - 
1. Smaller and larger dense layers - speed is affected but I believe the model learns more with large initial dense layers.
2. Smaller initial conv layer - 16 neurons speed up the process, however training loss takes more time decaying.
3. Larger initial conv layer - 256 - expected better performance at the cost of speed but no significant improvement was observed. In fact, not only the model was slow in training but it was also slower than most in terms of loss decay.
5. Smaller neurons in all layer, no max pooling - expected faster decay (Hypothesis from jump1) - learning time went up pretty much, but we saw better performance
4. Removing max pooling completely - significantly increased the training time, even with lesser number of neurons in each layer
6. Removing Max pooling with just the first two layers - allowing them to learn more... while reducing the size in the bottom layers... Better Performance.
6. Having just 32/16 neurons in all the layers - significant improvement in initial decay of loss with lesser time for training...
8. 64 vs 32 nuerons in first layer - 64 (with pool size 3) - 3 mins training time, loss at the end of three epochs - 7.96, for 32 (with pool size 2) - training time2.5 minutes, loss at 7.96 -  we'll stick with 32 neurons for now
9. no pooling for 1 vs 2 layers - 2 - same for above, 1 - training time 2 mins 20 secs, loss - 7.96
10. Adding another layer of 16 nuerons with max pooling - time - 2 mins, loss - 8.47
11. Removing a layer of 16 neurons - time - 2mins 15 sec , loss -  7.75
12. Adding a conv layer and removing a large dense layer - time 115 sec, loss 8.35
12. Keeping new conv layer and removing smaller dense layers - time 2 mins, loss 8.39
13. Adding more smaller dense layers - 
             with additional CONV layer
            (1024, 256, 128, 64, 32,16) - time 125 sec, loss - 8.28 ,   
            (1024, 256, 128, 64, 64, 32, 32, 16, 16) - time 4 min+, loss stagnant at 80, time reduced to 3 min, loss - 7.6
            after 6 epochs
            with CONV layers removed
            1024, 256, 128, 64, 32,16) - time - 3.5 mins, loss - 7.55,   
                removing larger dense layer
                   (128, 64, 32, 16) - time - 3mins, loss - 7.81
                   (1024, 256) - time -, loss - 3.5 mins, loss - 9.76
10. no pooling for 1 layer - 32 vs 64 neurons - 64 - time - 268 sec, loss - 7.20, 64, 32, 16... 7.9, time - 2.5 minutes
                                                32 - time 2 minutes, loss - 8
11. 64 32 16 16 1024 256 - time - 160 sec, loss - 7.5
12. 64 32 16 16 512 256 - time 150 sec, 7.67
12. 64 32 16 16 256 64 - time 135 sec, 7.88
12. 64 32 16 64 32 16 - time 140 sec, 7.51
12. 64 32 16 16 16 64 32 16 - time 241 sec, 8.02
12. 64 32 16 16 64 32 16 - time 225 sec, 7.148

64 32 16 16 64 32 16 - final model to be trained for 100 epochs then compared with jump 1 -  High Validation and testing loss observed
kernel size 5 - loss 3.589, testing loss - 8.5+
kernel size 3 (slower) loss - 3.89, testing loss - 5.83
kernel size 7 - faster but large testing loss

### Conclusion
Shallow models are fast but they fail at reducing the losses and learning beyond a certain point.


Best models so far -
1. jump1 - trained for more than a hundred epochs - 2.1 loss
2. jump5 - trained for 100 epochs - 3.5 loss, but testing loss>8
3. jump6 - trained for 100 epochs - 3.89 loss with testing loss >5

Optimizers tried
1. Adam
2. Adadelta
3. RMSProp

Best suited so far - adam
                  